qoqo/src/circuit.rs

```diff
@@ -602,74 +604,69 @@ impl CircuitWrapper {
-                            "Right hand side cannot be converted to Operation or Circuit {:?}",
-                            x
-                        ))
-                    });
-                    match other {
-                        Ok(x) => {
-                            self.internal += x;
-                            Ok(())
-                        }
-                        Err(y) => Err(y),
+    fn __iadd__(&mut self, other: &Bound<PyAny>) -> PyResult<()> {
+        match convert_pyany_to_operation(other) {
+            Ok(x) => {
+                self.internal += x;
+                Ok(())
+            }
+            Err(_) => {
+                let other = convert_into_circuit(other).map_err(|x| {
+                    pyo3::exceptions::PyTypeError::new_err(format!(
+                        "Right hand side cannot be converted to Operation or Circuit {:?}",
+                        x
+                    ))
+                });
+                match other {
+                    Ok(x) => {
+                        self.internal += x;
+                        Ok(())
                     }
+                    Err(y) => Err(y),
                 }
             }
-        })
+        }
     }
 
     /// Implement the `+` (__add__) magic method to add two Circuits.
     ///
     /// Args:
-    ///     lhs (Circuit): The first Circuit object in this operation.
+    ///     self (CircuitWrapper): The first Circuit object in this operation.
     ///     rhs (Circuit): The second Circuit object in this operation.
     ///
     /// Returns:
-    ///     lhs + rhs (Circuit): the two Circuits added together.
+    ///     self + rhs (Circuit): the two Circuits added together.
     ///
     /// Raises:
     ///     TypeError: Left hand side can not be converted to Circuit.
     ///     TypeError: Right hand side cannot be converted to Operation or Circuit.
-    fn __add__(lhs: Py<PyAny>, rhs: Py<PyAny>) -> PyResult<CircuitWrapper> {
-        Python::with_gil(|py| -> PyResult<CircuitWrapper> {
-            let (lhs_ref, rhs_ref) = (lhs.as_ref(py), rhs.as_ref(py));
-            let self_circ = convert_into_circuit(lhs_ref).map_err(|_| {
-                PyTypeError::new_err("Left hand side can not be converted to Circuit")
-            })?;
-            match convert_pyany_to_operation(rhs_ref) {
-                Ok(x) => Ok(CircuitWrapper {
-                    internal: self_circ + x,
-                }),
-                Err(_) => {
-                    let other = convert_into_circuit(rhs_ref).map_err(|_| {
-                        pyo3::exceptions::PyTypeError::new_err(
-                            "Right hand side cannot be converted to Operation or Circuit",
-                        )
-                    })?;
-                    Ok(CircuitWrapper {
-                        internal: self_circ + other,
-                    })
+    fn __add__(&mut self, other: &Bound<PyAny>) -> PyResult<CircuitWrapper> {
```

Comment:
Is this the correct way to implement these? Should we be using the `staticmethod` marker instead?

Reply:
I think so, static method are for methods that doesn't use an instance of the class for example `Circuit::current_version()` and on the [pyo3 doc](https://pyo3.rs/v0.21.2/class/protocols) it says that the signature should be `__add__(<self>, object) -> object` where self is an instance of the class

Reply:
Indeed! Good point